wlutil/wlutil.py

```diff
@@ -311,3 +321,144 @@ def copyImgFiles(img, files, direction):
+    linuxMaj = str(int(linuxHeaderVer) >> 16)
+    linuxMin = str((int(linuxHeaderVer) >> 8) & 0xFF)
+
+    # Toolchain major version
+    toolVerStr = sp.run(["riscv64-unknown-linux-gnu-gcc", "--version"],
+            universal_newlines=True, stdout=sp.PIPE).stdout
+    toolVer = toolVerStr[36]
+
+    return {'linuxMaj' : linuxMaj,
+            'linuxMin' : linuxMin,
+            'gcc' : toolVer}
+
+def checkGitStatus(submodule):
+    """Returns a dictionary representing the status of a git repo.
+
+    submodule: Path to the submodule to check. This check will be skipped if
+        the empty string is passed (the empty string is always uptodate)
+
+    Fields:
+    'sha'  : latest git commit hash (or "" if not initialized)
+    'dirty': boolean indicating whether there are uncommited changes (uninitialized repos are considered dirty)
+    'init' : boolean indiicating if the repository has been initialized
+    'rebuild' : A random number if the repo should be considered not up to date
+        for any reason (e.g. dirty==True or init==False). 0 otherwised.
+
+    This is primarily useful as an input to doit's config_changed() updtodate
+    helper which considers a workload not uptodate if some string or dictionary
+    has changed since the last time it ran. The 'sha' or 'rebuild' fields will
+    change if the repo has changed (or we can't tell if it's changed)."""
+
+    log = logging.getLogger()
+
+    if submodule == None:
+        return {
+                'sha' : "",
+                'dirty' : False,
+                "init" : False,
+                "rebuild" : ""
+                }
+
+    try:
+        repo = git.Repo(submodule)
+    except (git.InvalidGitRepositoryError, git.exc.NoSuchPathError):
+        # Submodule not initialized (or otherwise can't be read as a repo)
+        return {
+                'sha' : "",
+                'dirty' : True,
+                "init" : False,
+                "rebuild" : random.random()
+                }
+
+    status = {
+            'init' : True,
+            'sha' : repo.head.object.hexsha,
+            'dirty' : repo.is_dirty()
+            }
+    if repo.is_dirty():
+        # In the absense of a clever way to record changes, we must assume that
+        # a dirty repo has changed since the last time we built.
+        log.warn("Submodule: " + str(submodule) + " has uncommited changes. Any dependent workloads will be rebuilt")
+        status['rebuild'] = random.random()
+    else:
+        status['rebuild'] = 0
+
+    return status
+
+def checkSubmodule(s):
+    """Check whether a submodule is present and initialized.
+
+    It is safe to call this on something that is not actually a submodule. In
+    that case, this will simply check if the directory is empty or not.
+
+    s: Pathlib path to submodule
+
+    raises SubmoduleError if submodule not ready 
+    """
+    
+    if not s.exists() or not any(os.scandir(s)):
```

Comment:
This doesn't cover the case that a submodule is uninitialized but the directory is non-empty, but the subsequent `make` error is intuitive enough to figure out what went wrong.